navigator.cookieEnabled should return false when cookies are actually blocked

szewai opened the pull request:
#### 78ba6679d9abb85c6a70a713296eb03605092c20
<pre>
navigator.cookieEnabled should return false when cookies are actually blocked
<a href="https://bugs.webkit.org/show_bug.cgi?id=267789">https://bugs.webkit.org/show_bug.cgi?id=267789</a>
<a href="https://rdar.apple.com/121284878">rdar://121284878</a>

Reviewed by Chris Dumez.

According to <a href="https://html.spec.whatwg.org/multipage/system-state.html#dom-navigator-cookieenabled">https://html.spec.whatwg.org/multipage/system-state.html#dom-navigator-cookieenabled</a>, cookieEnabled should
return false when cookies cannot be set (i.e. cookies are blocked). However, in our current implementation, when cookies
are blocked, cookieEnabled may return true. This is because WebCookieJar::cookiesEnabled returns true when cookie accept
policy is not never, but WebKit does not rely on cookies accept policy to decide whether cookies should be blocked, see
NetworkStorageSession::shouldBlockCookies, where network process decides whether to block cookies.

Other cookies operations in WebCookieJar send sync messages to network process for answer. To make the result of
cookieEnabled consistent with the other operations, WebCookieJar::cookiesEnabled now also fetches the result from
network process. To avoid performance issues, instead of sending sync message, web process prefetches and cache the
result in Document. The cache is updated on demand, like when document navigates to another URL or when cookie blocking
policy changes.

* LayoutTests/http/tests/resourceLoadStatistics/exemptDomains/managed-domains-cookieEnabled-expected.txt: Added.
* LayoutTests/http/tests/resourceLoadStatistics/exemptDomains/managed-domains-cookieEnabled.html: Added.
* LayoutTests/http/tests/resourceLoadStatistics/resources/managed-domains-cookieEnabled-iframe.html: Added.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::setCookieURL):
(WebCore::Document::setFirstPartyForCookies):
(WebCore::Document::updateCachedCookiesEnabled):
* Source/WebCore/dom/Document.h:
(WebCore::Document::cachedCookiesEnabled const):
(WebCore::Document::setCachedCookiesEnabled):
(WebCore::Document::setFirstPartyForCookies): Deleted.
* Source/WebCore/loader/CookieJar.cpp:
(WebCore::CookieJar::cookiesEnabled):
(WebCore::CookieJar::remoteCookiesEnabled const):
(WebCore::CookieJar::cookiesEnabled const): Deleted.
* Source/WebCore/loader/CookieJar.h:
* Source/WebCore/platform/network/NetworkStorageSession.cpp:
(WebCore::NetworkStorageSession::setPrevalentDomainsToBlockAndDeleteCookiesFor):
(WebCore::NetworkStorageSession::setPrevalentDomainsToBlockButKeepCookiesFor):
(WebCore::NetworkStorageSession::setDomainsWithUserInteractionAsFirstParty):
(WebCore::NetworkStorageSession::setAppBoundDomains):
(WebCore::NetworkStorageSession::resetAppBoundDomains):
(WebCore::NetworkStorageSession::setManagedDomains):
(WebCore::NetworkStorageSession::resetManagedDomains):
(WebCore::NetworkStorageSession::cookiesEnabled const):
(WebCore::NetworkStorageSession::addCookiesEnabledStateObserver):
(WebCore::NetworkStorageSession::removeCookiesEnabledStateObserver):
(WebCore::NetworkStorageSession::cookieEnabledStateMayHaveChanged):
* Source/WebCore/platform/network/NetworkStorageSession.h:
(WebCore::CookiesEnabledStateObserver::~CookiesEnabledStateObserver):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::~NetworkConnectionToWebProcess):
(WebKit::NetworkConnectionToWebProcess::cookiesEnabledSync):
(WebKit::NetworkConnectionToWebProcess::cookiesEnabled):
(WebKit::NetworkConnectionToWebProcess::cookieEnabledStateMayHaveChanged):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
(WebKit::NetworkConnectionToWebProcess::protectedConnection):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.cpp:
(WebKit::NetworkProcessConnection::updateCachedCookiesEnabled):
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.h:
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.messages.in:
* Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
(WebKit::WebCookieJar::cookiesEnabled):
(WebKit::WebCookieJar::remoteCookiesEnabledSync const):
(WebKit::WebCookieJar::remoteCookiesEnabled const):
(WebKit::WebCookieJar::cookiesEnabled const): Deleted.
* Source/WebKit/WebProcess/WebPage/WebCookieJar.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::setThirdPartyCookieBlockingMode):
(WebKit::WebProcess::updateCachedCookiesEnabled):
* Source/WebKit/WebProcess/WebProcess.h:

Canonical link: <a href="https://commits.webkit.org/273522@main">https://commits.webkit.org/273522@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/18eb839da3492578e18d5457f9035a77e23551c8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/35656 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/14596 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/37795 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/38394 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/32118 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/36855 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/16986 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/11622 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/30903 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/36209 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/12341 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/31731 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/10842 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/10848 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/31872 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/39639 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/32404 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/32208 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/36802 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/11037 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/8928 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/34883 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/12755 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/31530 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8140 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/11545 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/11823 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
```diff
@@ -39,6 +39,7 @@ messages -> NetworkConnectionToWebProcess LegacyReceiver {
 
     CookiesForDOM(URL firstParty, struct WebCore::SameSiteInfo sameSiteInfo, URL url, WebCore::FrameIdentifier frameID, WebCore::PageIdentifier pageID, enum:bool WebCore::IncludeSecureCookies includeSecureCookies, enum:bool WebCore::ApplyTrackingPrevention applyTrackingPrevention, enum:bool WebCore::ShouldRelaxThirdPartyCookieBlocking shouldRelaxThirdPartyCookieBlocking) -> (String cookieString,  ...
     SetCookiesFromDOM(URL firstParty, struct WebCore::SameSiteInfo sameSiteInfo, URL url, WebCore::FrameIdentifier frameID, WebCore::PageIdentifier pageID, enum:bool WebCore::ApplyTrackingPrevention applyTrackingPrevention, String cookieString, enum:bool WebCore::ShouldRelaxThirdPartyCookieBlocking shouldRelaxThirdPartyCookieBlocking)
+    CookiesEnabled(URL firstParty, URL url, std::optional<WebCore::FrameIdentifier> frameID, std::optional<WebCore::PageIdentifier> pageID, enum:bool WebCore::ShouldRelaxThirdPartyCookieBlocking shouldRelaxThirdPartyCookieBlocking) -> (bool enabled) Synchronous
```
Don't introduce a sync IPC for this.

szewai replied:
Uploaded patch with new approach to prefetch and cache the value in `Document`.

cdumez commented on Source/WebCore/platform/network/NetworkStorageSession.h:
```diff
@@ -281,6 +291,7 @@ class NetworkStorageSession : public CanMakeWeakPtr<NetworkStorageSession> {
     RetainPtr<NSMutableSet> m_subscribedDomainsForCookieChanges;
     MemoryCompactRobinHoodHashMap<String, HashSet<CheckedPtr<CookieChangeObserver>>> m_cookieChangeObservers;
 #endif
+    HashSet<CheckedPtr<CookiesEnabledStateObserver>> m_cookiesEnabledStateObservers;
```
We only use CheckedPtr/CheckedRef for stack variables. Please use WeakRef<> here.

cdumez commented on Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
```diff
@@ -217,11 +217,29 @@ bool WebCookieJar::cookiesEnabled(const Document& document) const
+    if (applyTrackingPreventionInNetworkProcess == ApplyTrackingPrevention::No)
+        return true;
+
+    return document.cachedCookiesEnabled();
```
What if `cachedCookiesEnabled()` hasn't been initialized yet? Since it is updated asynchronously via async IPC, it seems like we may return an incorrect result temporarily, which may lead to flakiness.

szewai replied:
I've made cachedCookiesEnabled() return an optional, and added a check for it so it logs error when the value is not cached.
I am not sure whether it's a common case that cookiesEnabled is checked before it's cached, since cookieURL seems to be set very early in the loading stage. So currently if the value is not cached yet, we return true (note the subsequent cookie getting/setting will still fail as they will send message to network process).
If this error is popular, we might consider adding sync message for cookiesEnabled later. WDYT?

cdumez commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -6215,8 +6215,38 @@ void Document::setCookieURL(const URL& url)
+
+void Document::setCachedCookiesEnabled(bool enabled)
+{
+    m_cachedCookiesEnabled = enabled;
```
Please inline in header.

cdumez commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -6215,8 +6215,38 @@ void Document::setCookieURL(const URL& url)
+    }
+
+    page->cookieJar().cookiesEnabledAsync(*this, [weakThis = WeakPtr<Document, WeakPtrImplWithEventTargetData> { *this }](bool enabled) mutable {
+        if (weakThis)
```
`if (RefPtr document = weakDocument.get())`

cdumez commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -6215,8 +6215,38 @@ void Document::setCookieURL(const URL& url)
+
+    page->cookieJar().cookiesEnabledAsync(*this, [weakThis = WeakPtr<Document, WeakPtrImplWithEventTargetData> { *this }](bool enabled) mutable {
+        if (weakThis)
+            weakThis->setCachedCookiesEnabled(enabled);
```
`document->`

cdumez commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -6215,8 +6215,38 @@ void Document::setCookieURL(const URL& url)
+        return;
+    }
+
+    page->cookieJar().cookiesEnabledAsync(*this, [weakThis = WeakPtr<Document, WeakPtrImplWithEventTargetData> { *this }](bool enabled) mutable {
```
`weakDocument = WeakPtr { *this }` (no need for template parameters and renamed)

cdumez commented on Source/WebCore/dom/Document.h:
```diff
@@ -2540,6 +2543,7 @@ class Document
     RefPtr<ResizeObserver> m_resizeObserverForContainIntrinsicSize;
 
     const std::optional<FrameIdentifier> m_frameIdentifier;
+    bool m_cachedCookiesEnabled { true };
```
Odds are this should be a std::optional<> so that we know if the cache is valid or not.

cdumez commented on Source/WebCore/loader/CookieJar.cpp:
```diff
@@ -132,15 +132,29 @@ void CookieJar::setCookies(Document& document, const URL& url, const String& coo
 }
 
-bool CookieJar::cookiesEnabled(const Document&) const
+bool CookieJar::cookiesEnabled(const Document& document) const
```
Are the changes to this function really needed? Isn't this only used by WebKitLegacy?

szewai replied:
Yes, this is only used for WebKitLegacy. I made the change anyways since the code is still there, and result is supposed to be consistent.

cdumez commented on Source/WebCore/loader/CookieJar.cpp:
```diff
@@ -132,15 +132,29 @@ void CookieJar::setCookies(Document& document, const URL& url, const String& coo
 
+void CookieJar::cookiesEnabledAsync(const Document&, CompletionHandler<void(bool)>&& completionHandler) const
+{
+    completionHandler(false);
```
Why is it OK to always return false here? I think this is only used by WebKitLegacy but I am not entirely clear why we return false.

szewai replied:
This is also called in WebKit as we create place holder Document for SharedWorker (`NonRenderedPlaceholder`), and Document will try fetch the value anyways when cookieURL and firstPartyForCookies is set. 
cookieEnabled is only available in Navigator, not WorkerNavigator, so I made it return false.

cdumez commented on Source/WebCore/platform/network/NetworkStorageSession.h:
```diff
@@ -101,6 +101,12 @@ class CookieChangeObserver : public CanMakeCheckedPtr {
 };
 #endif
 
+class CookiesEnabledStateObserver : public CanMakeCheckedPtr {
```
`CanMakeWeakPtr< CookiesEnabledStateObserver>`

cdumez commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -2335,6 +2339,12 @@ void WebProcess::removeWebTransportSession(WebTransportSessionIdentifier identif
 
+void WebProcess::updateCachedCookiesEnabled()
+{
+    for (auto& document : Document::allDocuments())
```
`Ref`

szewai replied:
It turns out allDocuments()  returns CheckRef, and it gives compile error when using Ref here; probably because CheckedRef cannot be converted to Ref directly.

cdumez commented on Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
```diff
@@ -217,11 +218,34 @@ bool WebCookieJar::cookiesEnabled(const Document& document) const
+        return true;
+
+    if (!document.cachedCookiesEnabled()) {
+        RELEASE_LOG_ERROR(Network, "WebCookieJar::cookiesEnabled remote is not cached yet");
```
In my opinion, this is not any better than the previous iteration. It can still lead to returning an incorrect result.
I think we need to do a sync IPC in this particular edge case, to make sure we consistently return the right value and don't end up with flaky results.

cdumez commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
```diff
@@ -911,6 +932,11 @@ void NetworkConnectionToWebProcess::allCookiesDeleted()
 
+void NetworkConnectionToWebProcess::cookieEnabledStateMayHaveChanged()
+{
+    connection().send(Messages::NetworkProcessConnection::UpdateCachedCookiesEnabled(), 0);
```
Please add a `protectedConnection()` getter which returns a smart pointer and call it here.

szewai replied:
I find there are many places using `connection()` in this file, so let me land this patch first and I will do a follow-up to change all of them: https://bugs.webkit.org/show_bug.cgi?id=268085.

cdumez replied:
Please do the change I suggest in this patch. You don't have to fix ALL cases in this PR, just this one.

szewai replied:
Added.

cdumez replied:
Thank you.

cdumez reviewed (approved):
LGTM with nit.

szewai committed:
navigator.cookieEnabled should return false when cookies are actually blocked
https://bugs.webkit.org/show_bug.cgi?id=267789
rdar://121284878

Reviewed by Chris Dumez.

According to https://html.spec.whatwg.org/multipage/system-state.html#dom-navigator-cookieenabled, cookieEnabled should
return false when cookies cannot be set (i.e. cookies are blocked). However, in our current implementation, when cookies
are blocked, cookieEnabled may return true. This is because WebCookieJar::cookiesEnabled returns true when cookie accept
policy is not never, but WebKit does not rely on cookies accept policy to decide whether cookies should be blocked, see
NetworkStorageSession::shouldBlockCookies, where network process decides whether to block cookies.

Other cookies operations in WebCookieJar send sync messages to network process for answer. To make the result of
cookieEnabled consistent with the other operations, WebCookieJar::cookiesEnabled now also fetches the result from
network process. To avoid performance issues, instead of sending sync message, web process prefetches and cache the
result in Document. The cache is updated on demand, like when document navigates to another URL or when cookie blocking
policy changes.

* LayoutTests/http/tests/resourceLoadStatistics/exemptDomains/managed-domains-cookieEnabled-expected.txt: Added.
* LayoutTests/http/tests/resourceLoadStatistics/exemptDomains/managed-domains-cookieEnabled.html: Added.
* LayoutTests/http/tests/resourceLoadStatistics/resources/managed-domains-cookieEnabled-iframe.html: Added.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::setCookieURL):
(WebCore::Document::setFirstPartyForCookies):
(WebCore::Document::updateCachedCookiesEnabled):
* Source/WebCore/dom/Document.h:
(WebCore::Document::cachedCookiesEnabled const):
(WebCore::Document::setCachedCookiesEnabled):
(WebCore::Document::setFirstPartyForCookies): Deleted.
* Source/WebCore/loader/CookieJar.cpp:
(WebCore::CookieJar::cookiesEnabled):
(WebCore::CookieJar::remoteCookiesEnabled const):
(WebCore::CookieJar::cookiesEnabled const): Deleted.
* Source/WebCore/loader/CookieJar.h:
* Source/WebCore/platform/network/NetworkStorageSession.cpp:
(WebCore::NetworkStorageSession::setPrevalentDomainsToBlockAndDeleteCookiesFor):
(WebCore::NetworkStorageSession::setPrevalentDomainsToBlockButKeepCookiesFor):
(WebCore::NetworkStorageSession::setDomainsWithUserInteractionAsFirstParty):
(WebCore::NetworkStorageSession::setAppBoundDomains):
(WebCore::NetworkStorageSession::resetAppBoundDomains):
(WebCore::NetworkStorageSession::setManagedDomains):
(WebCore::NetworkStorageSession::resetManagedDomains):
(WebCore::NetworkStorageSession::cookiesEnabled const):
(WebCore::NetworkStorageSession::addCookiesEnabledStateObserver):
(WebCore::NetworkStorageSession::removeCookiesEnabledStateObserver):
(WebCore::NetworkStorageSession::cookieEnabledStateMayHaveChanged):
* Source/WebCore/platform/network/NetworkStorageSession.h:
(WebCore::CookiesEnabledStateObserver::~CookiesEnabledStateObserver):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::~NetworkConnectionToWebProcess):
(WebKit::NetworkConnectionToWebProcess::cookiesEnabledSync):
(WebKit::NetworkConnectionToWebProcess::cookiesEnabled):
(WebKit::NetworkConnectionToWebProcess::cookieEnabledStateMayHaveChanged):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
(WebKit::NetworkConnectionToWebProcess::protectedConnection):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.cpp:
(WebKit::NetworkProcessConnection::updateCachedCookiesEnabled):
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.h:
* Source/WebKit/WebProcess/Network/NetworkProcessConnection.messages.in:
* Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
(WebKit::WebCookieJar::cookiesEnabled):
(WebKit::WebCookieJar::remoteCookiesEnabledSync const):
(WebKit::WebCookieJar::remoteCookiesEnabled const):
(WebKit::WebCookieJar::cookiesEnabled const): Deleted.
* Source/WebKit/WebProcess/WebPage/WebCookieJar.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::setThirdPartyCookieBlockingMode):
(WebKit::WebProcess::updateCachedCookiesEnabled):
* Source/WebKit/WebProcess/WebProcess.h:

Canonical link: https://commits.webkit.org/273522@main
